files/private-chef-ctl-commands/key-control.rb

```diff
@@ -0,0 +1,255 @@
+      STDERR.puts "Usage: Default name used is the fingerprint of the key passed."
+      exit 1
+    end
+  end
+
+  def add_key(cmd_args, id)
+    # TODO should keyname be optional and default the key fingerprint?
+
+    # this is guaranteed to work because it would have crashed above if not
+    # expiration_date and key_name will be nil if they weren't passed
+    arg_parser = ArgParser.new
+    expiration_date = arg_parser.find_arg_after_identifier(cmd_args, "--expiration-date")
+    key_name        = arg_parser.find_arg_after_identifier(cmd_args, "--key-name")
+    user = cmd_args[0]
+    key = File.read(cmd_args[1])
+
+    unless /-----BEGIN PUBLIC KEY-----/.match(key)
+      STDERR.puts "Error: Invalid public key passed. Key must begin with:"
+      STDERR.puts "Error: -----BEGIN PUBLIC KEY-----"
+      exit 1
+    end
+
+    # if a key_name was not specified, use the fingerprint of the key
+    unless key_name
+      begin
+        openssl_key_object = OpenSSL::PKey::RSA.new(key)
+        data_string = OpenSSL::ASN1::Sequence([
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.n),
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.e)
+                                              ])
+        key_name = OpenSSL::Digest::SHA1.hexdigest(data_string.to_der).scan(/../).join(':')
+      rescue
+        STDERR.puts "Error: Could not parse fingerprint for public key #{cmd_args[1]}."
+        STDERR.puts "Error: If optional --key-name arg is not specified, then fingerprint of key is used by default as the key_id."
+        exit 1
+      end
+    end
+
+
+    if expiration_date
+      begin
+        Time.parse(expiration_date)
+      rescue
+        STDERR.puts "Error: Invalid date format inserted."
+        STDERR.puts "Error: You inserted #{expiration_date}."
+        STDERR.puts "Error: Please a date in format YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00."
+        exit 1
+      end
+    else
+      expiration_date = "infinity"
+    end
+
+    object = {
+      'id'          => id,
+      'key_id'      => key_name,
+      'public_key'  => key,
+      'key_version' => 0,
+      'created_at'  => Time.now.utc,
+      'expires_at'  => expiration_date
+    }
+
+    begin
+      DB[:keys].insert(object)
+    rescue Sequel::UniqueConstraintViolation
+      puts "Error: The key_name you passed (#{key_name}) was not unique for your object (#{cmd_args[0]})."
+      exit 1
+    end
+  end
+end
+
+add_command_under_category "add-client-key", "key-rotation", "Create a new client key", 2 do
+  addkey = AddKey.new
+  addkey.check_client_args(cmd_args, "Usage: chef-server-ctl add-client-key <orgname> <clientname> <path_to_public_key> [--expiration-date <date>] [--key-name <string>]")
+
+  id = nil
+  DB.fetch("SELECT id FROM clients WHERE name='#{cmd_args[1]}'") do |row|
```

Comment:
I think that you will also need the org_id in this query.  Otherwise the query could return multiple clients and the id you will get back will just be whatever the last one is.

Specifically, I think you want to do lines 154 through 161 first.  Then use the org_id in a query like:

```
DB.fetch("SELECT id FROM clients WHERE name='#{cmd_args[1]}' AND org_id='#{org_id}'")
```

or, using the sequel-y methods:

```
DB.select(:id).from(:clients).where(:name => cmd_args[1], :org_id => org_id)
```